Wrappers/Python/cil/optimisation/algorithms/SPDHG.py

```diff
@@ -46,69 +50,238 @@ class SPDHG(Algorithm):
-
+    sampler: an instance of a `cil.optimisation.utilities.Sampler` class
+        Method of selecting the next index for the SPDHG update. If None, random sampling and each index will have probability = 1/number of subsets
+        
     **kwargs:
+    prob : list of floats, optional, default=None
+        List of probabilities. If None each subset will have probability = 1/number of subsets. To be deprecated/ 
     norms : list of floats
-        precalculated list of norms of the operators
-
+        precalculated list of norms of the operators. To be deprecated and placed by the `set_norms` functionalist in a BlockOperator.
     Example 
     -------
 
     Example of usage: See https://github.com/vais-ral/CIL-Demos/blob/master/Tomography/Simulated/Single%20Channel/PDHG_vs_SPDHG.py
 
 
+
+    Note
+    -----
+    When setting `sigma` and `tau`, there are 4 possible cases considered by setup function: 
+
+    - Case 1: If neither `sigma` or `tau` are provided then `sigma` is set using the formula:
+        .. math:: 
+
+        \sigma_i=0.99 / (\|K_i\|**2)
+
+        and `tau` is set as per case 2
+
+    - Case 2: If `sigma` is provided but not `tau` then `tau` is calculated using the formula 
+
+        .. math:: 
+        \tau = 0.99\min_i([p_i / (\sigma_i * \|K_i\|**2) ])
+
+    - Case 3: If `tau` is provided but not `sigma` then `sigma` is calculated using the formula
+
+        .. math:: 
+        \sigma_i=0.99 p_i / (\tau*\|K_i\|**2)
+
+    - Case 4: Both `sigma` and `tau` are provided.
+
+
     Note
     ----
-    
+
     Convergence is guaranteed provided that [2, eq. (12)]:
-    
+
     .. math:: 
-    
     \|\sigma[i]^{1/2} * K[i] * tau^{1/2} \|^2  < p_i for all i
-    
-    Note
-    ----
-    
-    Notation for primal and dual step-sizes are reversed with comparison
-        to PDHG.py
-
-    Note
-    ----
 
-    this code implements serial sampling only, as presented in [2]
-        (to be extended to more general case of [1] as future work)             
-        
     References
     ----------
-    
+
     [1]"Stochastic primal-dual hybrid gradient algorithm with arbitrary 
     sampling and imaging applications",
     Chambolle, Antonin, Matthias J. Ehrhardt, Peter Richtárik, and Carola-Bibiane Schonlieb,
     SIAM Journal on Optimization 28, no. 4 (2018): 2783-2808.   
-        
+
     [2]"Faster PET reconstruction with non-smooth priors by randomization and preconditioning",
     Matthias J Ehrhardt, Pawel Markiewicz and Carola-Bibiane Schönlieb,
     Physics in Medicine & Biology, Volume 64, Number 22, 2019.
     '''
-    
-    def __init__(self, f=None, g=None, operator=None, tau=None, sigma=None,
-                 initial=None, prob=None, gamma=1.,**kwargs):
+
+    def __init__(self, f=None, g=None, operator=None, sigma=None, tau=None,
+                 initial=None, sampler=None,  **kwargs):
 
         super(SPDHG, self).__init__(**kwargs)
```

Comment:
Pull out all kwarg arguments to pass to algorithm. Then check the rest of them and give an error if there is any left